src/sparse/slicing.rs

```diff
@@ -0,0 +1,122 @@
+//! This module implementations to slice a matrix along the desired dimension.
+//! We're using a sealed trait to enable using ranges for an idiomatic API.
+
+use crate::range::Range;
+
+impl<N, I, Iptr> crate::CsMatI<N, I, Iptr>
```

Comment:
This duplication can be dropped if we implement all the methods on `CsMatBase`, but we would lose the opportunity to use `.view().slice_outer()` (lifetime problems). `.slice_outer()` would work directly

Reply:
And one would need a mutating slicer, `.slice_outer_mut()`. I'm not sure which approach is preferrable

Reply:
Yes this topic is a recurring issue, I've tried both approaches over the codebase, and I still can't settle on one approach. What's sure is that there needs to be a method on `CsMatViewI` that returns the lifetime of the type and not the lifetime of `&self`, otherwise some usages are incovenient or impossible.

In the past I've tried three approaches:

- implement on `CsMatBase`, and have another implementation on `CsMatViewI` with another name (e.g. with the suffix `_rbr` to mean the method "reborrows". The upside is that one is guaranteed to have a working implementation of the method, but the downside is that there is some duplication, and the reborrowing method is less discoverable (which can be accounted for through documentation)
- implement only on `CsMatViewI`, requiring to always call `.view().method()` for other types. It's the simplest approach in terms of implementation, with no duplication. But it suffers from a discoverability issue, greater than the previous one. I like this approach though, as it mirrors the relationship between `Vec<T>` and `&[T]`, but there the `Deref`trait makes the relationship seamless.
- Implement explicitly on all types, as here. The upside is discoverability I guess, but it arguably creates a more complex and duplicated implementation. I also risk missing some instatiations of `CsMatBase`.

That's my current state of mind. Ideally I'd like to take this PR as an opportunity to discuss these alternatives, pick one approach and try to be consistent throughout the crate for 0.10.

Reply:
For the third approach, it may be possible to reduce duplication using macros. Another downside I had not mentionned is bloat (but probably similar to implementing on `CsMatBase` because of monomorphization. Which means the second approach has the advantage of generating tighter code.

Reply:
The third option will lead to N entries on the left hand side of the [docs of CsMatBase](https://docs.rs/sprs/0.9.2/sprs/struct.CsMatBase.html), with different signatures. It would also exclude other items implementing `Deref<[N]>`, such as `Box<[u8]>`, `Cow<u8>` etc.

The first option might be the most consistent, and will prevent a peppering of `view()` all over the place. I also remember having the same sort of problems with temporary views with `ndarray`, so it shouldn't be surprising when such errors does emerge (well, equally suprising anyway).

The discoverability of the second option would not be too bad if we added a note to each applicable function. (this method requires a view, use `view()`)

Reply:
Why exactly do we need a `reborrow` version of the methods? The following would work:
```rust
let v = csr.view();
let m = v.foo();
```
Is it to keep `m` after `v` goes out of scope?

Reply:
Regarding monomorphization: We could always have internal methods on `CsMatView`/`CsMatViewMut` and call this from `CsMatBase`. Would be interesting to see the benefits in both compile time and crate size!

Reply:
> Why exactly do we need a `reborrow` version of the methods? The following would work:
> 
> ```rust
> let v = csr.view();
> let m = v.foo();
> ```
> 
> Is it to keep `m` after `v` goes out of scope?

There are some situations, like when implementing iterators, where the variable `v` would go out of scope before returning, but we know the data is still valid. See https://github.com/vbarrielle/sprs/blob/6cc2370f685c674886f958994f1d30123c6b0c42/src/sparse/csmat.rs#L1159 for an example.

Reply:
> Regarding monomorphization: We could always have internal methods on `CsMatView`/`CsMatViewMut` and call this from `CsMatBase`. Would be interesting to see the benefits in both compile time and crate size!

That's a good idea!

Reply:
With this argument about monomorphization, I think we're now down to two options:

- implement on `CsMatBase` a thin wrapper over the real implementation on `CsMatViewI`
- implement on `CsMatViewI` and add documentation comments for readability.

The third approach was the most complicated.

I guess the first approach is preferable, as it makes for the less surprising user experience. Would you agree?

Reply:
I agree!